src/utils.jl

```diff
@@ -107,3 +118,78 @@ end
   isnothing(ret) && return IteratorReturn(nothing)
   ret
 end
+
+# adapted from Base to handle world age
+function code_typed_by_type(@nospecialize(tt::Type);
+                            optimize::Bool=true,
+                            debuginfo::Symbol=:default,
+                            world::UInt=Base.get_world_counter(),
+                            interp::Core.Compiler.AbstractInterpreter=Core.Compiler.NativeInterpreter(world))
+    if isdefined(Base, :IRShow)
+        debuginfo = Base.IRShow.debuginfo(debuginfo)
+    elseif debuginfo === :default
+        debuginfo = :source
+    end
+    if debuginfo !== :source && debuginfo !== :none
+        throw(ArgumentError("'debuginfo' must be either :source or :none"))
+    end
+    tt = Base.to_tuple_type(tt)
+    min_world = Ref{UInt}(typemin(UInt))
+    max_world = Ref{UInt}(typemax(UInt))
+    match = ccall(:jl_gf_invoke_lookup_worlds, Any,
+              (Any, Any, Csize_t, Ref{Csize_t}, Ref{Csize_t}),
+              tt, #=mt=# nothing, world, min_world, max_world)
+    meth = Base.func_for_method_checked(match.method, tt, match.sparams)
+    (code, ty) = Core.Compiler.typeinf_code(interp, meth, match.spec_types, match.sparams, optimize)
+    if code === nothing
+        ast = meth
+    else
+        debuginfo === :none && Base.remove_linenums!(code)
+        ast = code
+    end
+    mi = ccall(:jl_specializations_get_linfo, Ref{Core.MethodInstance},
+            (Any, Any, Any), match.method, match.spec_types, match.sparams)
+    return mi, ast, min_world[], max_world[]
+end
+
+function fsmi_generator(world::UInt, source::LineNumberNode, passtype, fsmitype::Type{Type{T}}, fargtypes) where T
+    @nospecialize
+    tt = Base.to_tuple_type(fargtypes)
+    mi, ci, min_world, max_world = code_typed_by_type(tt; world, optimize=false)
```

Comment:
Are there any special considerations for returning an "appropriate error" from this context? I see in CassetteOverlay you actually return code that would raise the error instead of raising an error from the generator.